### PR TITLE
process: invokes setImmediate on uv check phase

### PR DESCRIPTION
**Pull request by legendecas**
- [x] `npm test` passes
- [x] tests and/or benchmarks are included

Also enables uv check/prepare handles in libtuv.

### CODE REVIEW

File: src/modules/iotjs_module_process.c [hunk 1]
```diff
@@ -111,6 +111,23 @@ static jerry_value_t WrapEval(const char* name, size_t name_len, char* source,
...
 }
 
+static uv_check_t check_handle;
+static void onUVCheckCallback(uv_check_t* handle) {
+  const jerry_value_t process = iotjs_module_get("process");
+  jerry_value_t jonunuvcheck =
+      iotjs_jval_get_property(process, IOTJS_MAGIC_STRING__ONUVCHECK);
+  IOTJS_ASSERT(jerry_value_is_function(jonunuvcheck));
+
+  jerry_value_t jres = jerry_call_function(jonunuvcheck, process, NULL, 0);
```
Comment (lolBig) [inline src/modules/iotjs_module_process.c]: use iotjs_make_callback to call js function to trigger nextTick

File: src/js/iotjs.js [hunk 2]
```diff
@@ -480,9 +480,29 @@
...
-    // TODO(Yorkie): use nextTick for now...
-    process.nextTick(callback);
+    if (typeof callback !== 'function') {
+      throw new Error('Expect a function on setImmediate');
+    }
+    if (process.__immediateCallbacks == null) {
+      process.__immediateCallbacks = [];
+      process._startUVCheck();
+    }
+    process.__immediateCallbacks.push(callback);
```
Comment (yorkie) [inline src/js/iotjs.js]: Could we optimize this access operation by cache this variable `immediateCallbacks` in outside scope and make it always be an array?

File: src/modules/iotjs_module_process.c [hunk 3]
```diff
@@ -111,6 +111,17 @@ static jerry_value_t WrapEval(const char* name, size_t name_len, char* source,
   return res;
 }
 
+static uv_check_t check_handle;
+static void onUVCheckCallback(uv_check_t* handle) {
+  const jerry_value_t process = iotjs_module_get("process");
+  jerry_value_t jonunuvcheck =
```
Comment (yorkie) [inline src/modules/iotjs_module_process.c]: This might be a typo? Seems `jonuvcheck`?

File: src/modules/iotjs_module_process.c [hunk 4]
```diff
@@ -111,6 +111,17 @@ static jerry_value_t WrapEval(const char* name, size_t name_len, char* source,
...
+      iotjs_jval_get_property(process, IOTJS_MAGIC_STRING__ONUVCHECK);
+  IOTJS_ASSERT(jerry_value_is_function(jonunuvcheck));
+
+  iotjs_make_callback(jonunuvcheck, process, iotjs_jargs_get_empty());
+
```
Comment (yorkie) [inline src/modules/iotjs_module_process.c]: No need to add a blank?

File: src/modules/iotjs_module_process.c [hunk 5]
```diff
@@ -464,6 +475,30 @@ JS_FUNCTION(ForceGC) {
   return jerry_create_boolean(true);
 }
 
+JS_FUNCTION(StartUVCheck) {
+  int status;
+  if (check_handle.data != NULL) {
+    status = uv_check_start(&check_handle, onUVCheckCallback);
```
Comment (yorkie) [inline src/modules/iotjs_module_process.c]: Suggestion function name: `UVCheckCallback`, the `on` is duplicated with `callback`.

File: src/modules/iotjs_module_process.c [hunk 6]
```diff
@@ -464,6 +475,30 @@ JS_FUNCTION(ForceGC) {
...
+    return jerry_create_number(status);
+  }
+  iotjs_environment_t* iotjs_env = iotjs_environment_get();
+  uv_loop_t* iotjs_loop = iotjs_environment_loop(iotjs_env);
+  status = uv_check_init(iotjs_loop, &check_handle);
+  IOTJS_ASSERT(status == 0);
+
+  check_handle.data = (void*)(uintptr_t) true;
```
Comment (yorkie) [inline src/modules/iotjs_module_process.c]: What's the line for use?

File: test/run_pass/test_process_setimmediate.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+'use strict';
+var assert = require('assert');
+var fs = require('fs');
+
+var expectedResult = [0, 1, 2]
+var result = []
```
Comment (yorkie) [inline test/run_pass/test_process_setimmediate.js]: The above 2 lines lacks semi-colons :)

**Comment by yorkie** [comment]
Labeled with `minor`, too.

File: test/run_pass/test_process_setimmediate.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
+'use strict';
+var assert = require('assert');
+var fs = require('fs');
+
+var expectedResult = [0, 1, 2]
```
Comment (lolBig) [inline test/run_pass/test_process_setimmediate.js]: missing semicolon

File: src/js/iotjs.js [hunk 9]
```diff
@@ -480,9 +480,29 @@
   process._onNextTick = next_tick._onNextTick;
 
   global.setImmediate = setImmediate;
+  process._onUVCheck = function () {
+    var callbacks = process.__immediateCallbacks;
+    process.__immediateCallbacks = null;
+    callbacks.forEach(it => {
```
[the remark this answers is not in the document]
Comment (legendecas) [reply src/js/iotjs.js]: May I ask you putting a suggestion on the thread?

[the remark this answers is not in the document]
Comment (yorkie) [reply src/js/iotjs.js]:
```suggestion
    callbacks.forEach(function () {
```

[the remark this answers is not in the document]
Comment (yorkie) [reply src/js/iotjs.js]:
```suggestion
    callbacks.forEach(function (it) {
```

[the remark this answers is not in the document]
Comment (legendecas) [reply src/js/iotjs.js]: Doesn't seem necessary to me, do you have any other objections against the anonymous one?

**Comment by yorkie** [comment]
Ping @legendecas, still leaves a https://github.com/Rokid/ShadowNode/pull/406#discussion_r231791609 from @lolBig.

File: src/js/iotjs.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (yorkie) [reply src/js/iotjs.js]: It doesn't use `this` inside this function or arrow function, so there is no need to use it.

[the remark this answers is not in the document]
Comment (legendecas) [reply src/js/iotjs.js]: Yet arrow function looks much more simpler if both two act same here, are there any other strong objection to switch to function expression?

[the remark this answers is not in the document]
Comment (yorkie) [reply src/js/iotjs.js]: After searching the source code with `() =>`, found we have such many lines are using arrow function, let's start using arrow function in core, but with a little nit, we should follow the [`arrow-parens`](https://github.com/Rokid/ShadowNode/blob/master/.eslintrc.js#L37) rule.

**Comment by yorkie** [comment]
Removed the label `wont-land-on-v0.11.x` because #417 fixed.

**Pull request merged**